### ISSUE TITLE
MSP over Crossfire CRSF protocol

### ISSUE THREAD

**Issue by ZinhoRacing** [iNavFlight/inav#3886]
## Current Behavior
Not implemented yet. Works fine in Betaflight

## Desired Behavior
Add capability to Inav because INAV is the best :)

## Suggested Solution
I copyed appropriate code to Inav and made custom exe for MatekF405 wing target, working fine.
So copy and paste from Betaflight.(not so easy as there is few files to update and parameters, function to update).

## Who does this impact? Who is this for?
Every one that run Inav + crossfire and openTX.

## Additional context
Works fine with Inav 2.0 custom build + Matek F405Wing + Crossfire + Horus X10- OpenTx2.2.2
I managed to update a nice Widget (FrskyTelemetryScript from Yapuu) and convert it to use CRSF protocol. Picture of the original Widget below
Also work with other scripts like betaflight-tx-lua-scripts, so you can setup VTX, PID etc.. over radio.

![image](https://user-images.githubusercontent.com/41761785/46011129-3a0fbb80-c0c5-11e8-87ff-7aab7f8c4154.png)
Looks nice on the radio screen and very nice to have when flying without FPV gear (ground station).

**Comment by wx4cb** [comment]
any news on this ?

**Comment by ZinhoRacing** [comment]
It is working fine on my custom build. I did the mod as well for version 2.0.1
I don't know how to make pull request.

**Comment by wx4cb** [comment]
@digitalentity should be able to do that

**Comment by shellixyz** [comment]
@ZinhoRacing If you can't make a PR please post the files here we'll take care of it

**Comment by teckel12** [comment]
@ZinhoRacing As I've been working with you on this, I'll be more than happy to help. Send me your changes and I'll take care of the PR.

**Comment by ZinhoRacing** [comment]
@teckel12 , Hi, sorry had no time to clean and post the files.
But I'll be more than happy if you can do the PR. Thank you.
@shellixyz I'll do thanks and I'll need your point of view as well as I modified the osd.c/osd.h file to avoid duplicate code for CRSF telem FM (fly modes) and save some space (remove the static of Fly mode function to be accessible by telem/crsf.c code). I merged Teckel12 code (flight modes).

I should be able to work on it this week.

**Comment by ZinhoRacing** [comment]
@teckel12, @shellixyz 

Here is the files I modified to add MSP over CRSF protocol and to have all needed informations for  telemetry scripts like Lua telemetry or Yapuu widget (on OpenTX radio), thanks to you Tecekl12 for the MSP telem idea. 
Tested OK on my configuration that is : 
Horus X10S / Open TX 2.2.2 / Inav 2.0.0 and 2.0.1 custom builds.

Minimum Files to update to have MSP over CRSF protocol working: 
- fc/config.h
- rx/crsf.h
- rx/crsf.c
- telemetry/crsf.h
- telemetry/crsf.c
- telemetry/msp_shared.h
All the modifications of that files are merge from betaflight code.

Additional files for telemetry scripts (perhaps will need a new issue): 
- fc/fc_msp.c
- msp/msp_protocol.h
- io/osd.h
- io/osd.c

I hope it is clear, don't hesitate to ask me more explanation if needed.
Thanks.
[fc.zip](https://github.com/iNavFlight/inav/files/2646323/fc.zip)
[msp_protocol.zip](https://github.com/iNavFlight/inav/files/2646324/msp_protocol.zip)
[rx.zip](https://github.com/iNavFlight/inav/files/2646325/rx.zip)
[telemetry.zip](https://github.com/iNavFlight/inav/files/2646328/telemetry.zip)
[io.zip](https://github.com/iNavFlight/inav/files/2646332/io.zip)

**Comment by teckel12** [comment]
@ZinhoRacing I'll put together a PR for the MSP over Crossfire. The additional stuff for telemetry may not even be required once Horus and Crossfire support is added to LuaTelemetry, see:

![inavhorus](https://user-images.githubusercontent.com/2592128/49485558-6d7b6e80-f809-11e8-846a-518195019b84.png)

**Comment by teckel12** [comment]
@ZinhoRacing Sorry, your "Minimum Files" code wouldn't compile against the development branch without errors (which are a no-go).  Are you getting **ANY** compile errors?  If you are, they need to be resolved.  Even if it creates the HEX, it's not acceptable for merging if there's errors.

The files I changed were:

- fc/config.h
- rx/crsf.h
- rx/crsf.c
- telemetry/crsf.h
- telemetry/crsf.c
- telemetry/msp_shared.h

And here's the compile output:
```
$ make TARGET=MATEKF722SE
settings.yaml -> settings_generated.h, settings_generated.c
%% debug.c
%% config_eeprom.c
%% pwm_output.c
%% cli.c
%% config.c
%% controlrate_profile.c
%% fc_core.c
./src/main/fc/fc_core.c: In function 'filterRc':
./src/main/fc/fc_core.c:647:57: warning: implicit declaration of function 'getLooptime' [-Wimplicit-function-declaration]
         biquadFilterInitLPF(&filteredCycleTimeState, 1, getLooptime());
                                                         ^~~~~~~~~~~
%% fc_init.c
%% fc_tasks.c
./src/main/fc/fc_tasks.c: In function 'fcTasksInit':
./src/main/fc/fc_tasks.c:266:34: warning: implicit declaration of function 'getLooptime' [-Wimplicit-function-declaration]
     rescheduleTask(TASK_GYROPID, getLooptime());
                                  ^~~~~~~~~~~
%% fc_msp.c
%% fc_msp_box.c
%% rc_adjustments.c
%% rc_controls.c
%% rc_modes.c
%% settings.c
%% stats.c
%% failsafe.c
%% imu.c
%% mixer.c
%% pid.c
./src/main/flight/pid.c: In function 'pidInitFilters':
./src/main/flight/pid.c:219:34: warning: implicit declaration of function 'getLooptime' [-Wimplicit-function-declaration]
     const uint32_t refreshRate = getLooptime();
                                  ^~~~~~~~~~~
%% pid_autotune.c
%% rth_estimator.c
%% servos.c
%% wind_estimator.c
%% beeper.c
%% lights.c
%% pwmdriver_i2c.c
%% pwm.c
%% rx.c
%% spektrum.c
%% eleres.c
%% acceleration.c
%% battery.c
%% compass.c
%% diagnostics.c
%% gyro.c
./src/main/sensors/gyro.c: In function 'gyroInitFilters':
./src/main/sensors/gyro.c:313:94: warning: implicit declaration of function 'getLooptime' [-Wimplicit-function-declaration]
             biquadRCFIR2FilterInit(stage2Filter[axis], gyroConfig()->gyro_stage2_lowpass_hz, getLooptime());
                                                                                              ^~~~~~~~~~~
%% initialisation.c
./src/main/sensors/initialisation.c: In function 'sensorsAutodetect':
./src/main/sensors/initialisation.c:54:13: warning: implicit declaration of function 'getLooptime' [-Wimplicit-function-declaration]
     accInit(getLooptime());
             ^~~~~~~~~~~
%% blackbox.c
./src/main/blackbox/blackbox.c: In function 'blackboxWriteSysinfo':
./src/main/blackbox/blackbox.c:1566:61: warning: implicit declaration of function 'getLooptime' [-Wimplicit-function-declaration]
         BLACKBOX_PRINT_HEADER_LINE("looptime", "%d",                        getLooptime());
                                                             ^~~~~~~~~~~
%% cms.c
%% cms_menu_battery.c
%% cms_menu_blackbox.c
%% cms_menu_imu.c
%% cms_menu_ledstrip.c
%% cms_menu_misc.c
%% cms_menu_navigation.c
%% cms_menu_osd.c
%% cms_menu_vtx_smartaudio.c
%% cms_menu_vtx_tramp.c
%% dashboard.c
%% gps.c
%% gps_ublox.c
%% gps_nmea.c
%% gps_naza.c
%% osd.c
%% navigation.c
%% navigation_fixedwing.c
%% navigation_fw_launch.c
%% navigation_geo.c
%% navigation_multicopter.c
%% navigation_pos_estimator.c
%% pitotmeter.c
%% rangefinder.c
%% opflow.c
%% crsf.c
%% frsky_d.c
%% ibus_shared.c
%% ibus.c
%% ltm.c
%% mavlink.c
%% smartport.c
%% telemetry.c
%% vtx.c
%% vtx_control.c
%% serial_softserial.c
Linking MATEKF722SE
Memory region         Used Size  Region Size  %age Used
           FLASH:         692 B        16 KB      4.22%
    FLASH_CONFIG:          0 GB        16 KB      0.00%
          FLASH1:      279280 B       480 KB     56.82%
             TCM:        7764 B        64 KB     11.85%
             RAM:       51676 B       192 KB     26.28%
       MEMORY_B1:          0 GB         0 GB     -1.#J%
arm-none-eabi-size ./obj/main/inav_MATEKF722SE.elf
   text    data     bss     dec     hex filename
 270712    9260   50176  330148   509a4 ./obj/main/inav_MATEKF722SE.elf
arm-none-eabi-objcopy -O ihex --set-start 0x8000000 obj/main/inav_MATEKF722SE.elf obj/inav_2.1.0_MATEKF722SE.hex
```

Maybe you were not using the `development` branch as the base?  That's critical to be merged into the next release.  Maybe you ignored the errors as it still "works"?  In any case, I can't submit a PR if there's errors. I messed with it for a while but I didn't have time to do any further diagnosis as I was expecting more of an error-free compile.

BTW, the compile worked without errors before I added your files.  Maybe you've merged some of the telemetry OSD stuff in these files?

**Comment by ZinhoRacing** [comment]
Hi @teckel12,
You right, I was not using the Dev branch sorry for that. I did the merge in my copy of 2.0.1 official code :(
I have no warning or errors that's why I posted the files (of course I would not post code with errors or warning).
I'll do it on Dev branch tonight and test it.
It was not supposed to give you lot of work, sorry again. Now I know I need to work on dev branch if I want a change to be merged.

About LuaTelemetry it looks nice !!! I'll install it on my radio, I like the views.

**Comment by teckel12** [comment]
It's totally fine, I was expecting a bit of work anyway.  Sometimes, you won't see errors if the files with errors are already compiled.  Try doing a clean build and see if you get any errors.  For example:

```
make clean TARGET=NAME
````

I'll look at it again, I just didn't have time last night.

As far as [Lua Telemetry](https://github.com/iNavFlight/LuaTelemetry) goes, the Horus view is working in the [`horus`](https://github.com/iNavFlight/LuaTelemetry/tree/horus) branch.  Crossfire support is still in the works, as it won't be possible till INAV 2.1 is released and the additional flight modes are added.

**Comment by ZinhoRacing** [comment]
Ok, job done.
I worked on Dev branch (2.1.0).
Compilation is fine without errors. (I did make clean first)
I flashed my matekF405SE (wing), I had to do some settings adjustment for new functionnalities (like braking stuf etc) as I had invalide setting flag ON, but after that all working fine with MSP over CRSF (tested over a script that allow to change PID / VTX setting with MSP commands).
OSD files change is to avoid duplicate code for flight mode (using the same code in OSD and CRSF FM).

So files to update (only MSP over CRSF related) :  
- fc/config.h
- io/osd.c
- io/osd.h
- rx/crsf.h
- rx/crsf.c
- telemetry/crsf.h
- telemetry/crsf.c

[telemetry.zip](https://github.com/iNavFlight/inav/files/2650895/telemetry.zip)
[rx.zip](https://github.com/iNavFlight/inav/files/2650892/rx.zip)
[fc.zip](https://github.com/iNavFlight/inav/files/2650887/fc.zip)
[io.zip](https://github.com/iNavFlight/inav/files/2650888/io.zip)

**Comment by ZinhoRacing** [comment]
Ohhh i missed one file : 
- telemetry/msp_shared.h

[msp_shared.zip](https://github.com/iNavFlight/inav/files/2650943/msp_shared.zip)

**Comment by ZinhoRacing** [comment]
And here is 2 screen shoot (pictures took with my cell phone) of my radio (HorusX10) with LUA widget that use MSP data over CRSF link.
One is displaying some data Inav configurator style.
Other one is telemetry screen.

![img_1290](https://user-images.githubusercontent.com/41761785/49550769-0d5df880-f8ed-11e8-8a7a-c014e10be02b.jpg)
![img_1291](https://user-images.githubusercontent.com/41761785/49551047-02579800-f8ee-11e8-934e-139e57849ec3.jpg)

I think I'll want a merge of LuaTelemetry widget and my @yaapu custom widget :)

**Comment by wx4cb** [comment]
oooh now THATs purdy.... 

wonder if that could be run on a linux machine 😉 😉

**Comment by teckel12** [comment]
@ZinhoRacing The problem with the OSD stuff for flight modes is that it then doesn't work with the iNav telemetry script. My goal is to get Crossfire working with extended flight modes (which has already been merged into development) and MSP over Crossfire so VTx control (and other features) can be used. If this breaks what I've done so far and makes iNav's telemetry script not function, I'll need to rework it so it works with already established programs.

**Comment by teckel12** [comment]
@ZinhoRacing It seems the `crsfFrameDeviceInfo` function sends a lot of data without any real value.  I'm concerned about sending too much information and slowing down the communication so I removed that.

I also removed the OSD code because doing it that way could slow things down more than speed them up or save memory.

**Comment by teckel12** [comment]
@ZinhoRacing There's also a lot of code inside `#if defined(USE_CRSF_CMS_TELEMETRY)` pre-compile conditionals, yet `USE_CRSF_CMS_TELEMETRY` is not defined so none of it is actually compiling.  I believe this code has to do with configuring settings via the OSD over Crossfire.  But, this is never turned on anyway so it's also been deleted.

I believe you mean to say that you use the MATEKF405SE target?  I've compiled it and you can test it here: 
[inav_2.1.0_MATEKF405SE.hex.zip](https://github.com/iNavFlight/inav/files/2651714/inav_2.1.0_MATEKF405SE.hex.zip)
If this is the wrong target, let me know the target name and I'll compile it.

It's probably best to test  the MSP with the betaflight-tx-lua-scripts.  Keep in mind that the device info stuff was removed so we shouldn't expect that with this hex file.

**Comment by ZinhoRacing** [comment]
The crsfFrameDeviceInfo function only send frame if requested by slave (master is FC and radio is slave). I tested it and it is working fine. It is part of native CRSF protocol (TBS). So I would be happy to leave it here.
About USE_CRSF_CMS_TELEMETRY code, I left it here because it is not compiled as you said and it is comming from the betaflight file. I did more or less a merge/copy of betaflight code. I agree that can be removed.
About OSD code ... I don't understand why it could slow down things ... and doing it that way we are sure to always have the same Inav Flight mode on OSD and on Inav CRSF telemetry if there is change or update of Inav flight modes. I would like to leave it.
I'll test the Hex tonight. Thanks.
Yes I have a MATEKF405 wing (target  = MATEKF405SE).
I'm using betaflight-tx-lua-scripts to test.

**Comment by teckel12** [comment]
@ZinhoRacing I'll look into the `crsfFrameDeviceInfo` further, thanks.

If a user doesn't use an OSD it will need to run it anyway. OSD is also a resource pig, so avoiding it if not needed would be beneficial. The way to do it would be an external file that would be included by both OSD and Crossfire, but, there are slight differences, like spacing so maybe keeping them different isn't such a bad idea. If another developer objects, we could do the new included file method.

**Comment by ZinhoRacing** [comment]
I like the external file idea and perhaps it could be used by the other telemetry protocols that expose Inav Flight modes. I didn't thought of non OSD boards, my apologizes ...
I flashed your Hex and MSP over CRSF is working just fine.
Tested it with BF scripts, changing PID values and VTX values. Reading GPS stats as well over MSP. All OK.
Nice job @teckel12

**Comment by teckel12** [comment]
@ZinhoRacing I believe Crossfire is the only telemetry that outputs a string, so it's not useful to other telemetry. There's also spacing differences and other slight differences that could throw off a legacy script.

It's not much of a big deal to have it in two places, there's virtually no computation involved. Now when you wanted to send all of the reasons why it wasn't able to arm, that was a case where having unified code would be beneficial.

Anyway, good to hear the code worked. I'll look again at the `crsfFrameDeviceInfo` function and do some testing locally with my Crossfire setup and probably submit the PR this weekend.

**Comment by teckel12** [comment]
@ZinhoRacing Here's the build with `crsfFrameDeviceInfo` added back in:

[inav_2.1.0_MATEKF405.hex.zip](https://github.com/iNavFlight/inav/files/2687939/inav_2.1.0_MATEKF405.hex.zip)
[inav_2.1.0_MATEKF405SE.hex.zip](https://github.com/iNavFlight/inav/files/2687940/inav_2.1.0_MATEKF405SE.hex.zip)

Could you post the LUA widget that you're using to show MSP data over CRSF link to display the data INAV configurator style?  I'd like to also test it with my Crossfire development build.  Once we can confirm it works, I can submit it as a pull request to hopefully get it included in INAV 2.1.0

**Comment by ZinhoRacing** [comment]
@teckel12 
Yes, no problem I'll clean the widget tonight and post here with explanation needed to install it on Horus.
It is all in the same widget. Inav data are displayed when selecting page 5. If you want to see how telemetry is working you need to change a file in INAV. (fc_msp.c) to include the special MSP for it MSP 155. not necessary to test only page 5 (INAV info configurator style). I'll test it with your build.

**Comment by ZinhoRacing** [comment]
Build is working fine regarding MSP functionality (MATEKF405SE).
I'm still working on my widget to cope with missing MSP messages ....

**Comment by teckel12** [comment]
I made some slight flight mode changes to the latest build.

1) `WAIT` doesn't show if extra arming safety is set to off (as `WAIT` is designed to indicate that it's waiting for a GPS lock and it should never wait for a GPS lock if extra arming safety is off)
1) New flight mode `AH` for altitude hold (altitude hold assumes ANGLE mode)
1) New flight mode `HRST` for home reset, which really isn't a flight mode, but indicates the setting of a new home position (this is a temporary flight mode, just while the home reset trigger is set)

I'll post new HEX files in a little bit.

**Comment by teckel12** [comment]
[inav_2.1.0_MATEKF405.hex.zip](https://github.com/iNavFlight/inav/files/2696924/inav_2.1.0_MATEKF405.hex.zip)
[inav_2.1.0_MATEKF405SE.hex.zip](https://github.com/iNavFlight/inav/files/2696925/inav_2.1.0_MATEKF405SE.hex.zip)
[inav_2.1.0_OMNIBUSF4PRO.hex.zip](https://github.com/iNavFlight/inav/files/2696926/inav_2.1.0_OMNIBUSF4PRO.hex.zip)

**Comment by wx4cb** [comment]
New flight mode AH for altitude hold (altitude hold assumes ANGLE mode) <<< wouldn't that be cruise mode if it also includes angle mode ?

Getting ready to go flying friday (tomorrow has a squall line coming thru FL), can i use a dump from 2.0.1 straight into 2.1 or are there things I would need to do in the conflagurator

**Comment by teckel12** [comment]
@wx4cb I don't do fixed wing so maybe? On multi-rotor, when I set alt hold it didn't say cruise (just stayed angle) which I why I added `AH`. 

Also, it's the development build for INAV 2.1 which is going to be a release candidate any day now. So I'd say it's as safe to fly as any pre-release version.

**Comment by teckel12** [comment]
@wx4cb Cruise mode with alt hold is `3CRS`, if not in cruise mode, alt hold would be `AH` which also activates angle mode. I believe the logic is sound. I've tested what I can (the cruise mode is FW only so I couldn't test that). But, the code is basically pulled from the OSD, so it's mostly well tested.

**Comment by wx4cb** [comment]
@teckel12 what I meant tim was that "cruise mode" (from memory) is alt hold with angle mode so they are effectively the same thing are they not ?

**Comment by teckel12** [comment]
@wx4cb Cruise mode is fixed wing, multi-rotor doesn't have cruise mode.  When I activate alt hold on multi-rotor, `AH` is displayed. There's never a cruise mode on multi-rotor. If there was (there's been some talk about adding it), it would show as cruise or `3CRS` just like the OSD. 

For fixed wing, there's cruise and cruise with alt hold, which is `3CRS`. Also, cruise with alt hold is first in the logic, so it takes presidence before `AH`.

Basically, there's no conflict with cruise mode.

**Comment by shellixyz** [comment]
@wx4cb CRUISE mode is not the same as AH + ANGLE. CRUISE mode is a specific mode for fixed wing that keeps the heading.

**Comment by wx4cb** [comment]
yea i realise that now. it was just my understanding that was wrong lol

**Comment by ZinhoRacing** [comment]
@teckel12 Hi,
finaly got it sorted ! was a problem with constant declaration ... bit strange LUA thing.
I tested it with your last Hex (the one with new flight modes AH - HRST).

so there is 2 zip files, one to put in WIDGETS folder and the other one in SCRIPTS folder.
[WIDGETS_Yaapu.zip](https://github.com/iNavFlight/inav/files/2700962/WIDGETS_Yaapu.zip)
[SCRIPTS_YAAPU.zip](https://github.com/iNavFlight/inav/files/2700963/SCRIPTS_YAAPU.zip)

To install it, just go to widgets and select Yaapu.
Then selecting Yaapu again you will be able to select the page you want with menu Widget Settings (page 5 is Inav infos, page 1 is telem page, few things not working with your build as MSP 155 and 156 are missing and I modified a bit 88 (VTX info, I add current power, band and channel to see real values  when vtx is powered up).  You can have 3 pages with Yaapu set to page 1, 5 and 2. page 2 is log messages.
Refresh is slow because it will try to get MSP 155 and 156 with 1s timeout for each and then will get the other MSP messages, cycling to them.

But you will be able to see how I use MSP over CRSF.

Page 1 
![image](https://user-images.githubusercontent.com/41761785/50316727-30c19f80-04b8-11e9-981b-92c49a373d05.png)

Page 5
![image](https://user-images.githubusercontent.com/41761785/50316735-420aac00-04b8-11e9-80f2-e76c3b6e43e7.png)

Page 2
![image](https://user-images.githubusercontent.com/41761785/50316757-55b61280-04b8-11e9-8d14-225e129b9826.png)

**Comment by ZinhoRacing** [comment]
Ho i Forget. To setup the widget. You need to go to scripts/yaapu and use menu.lua script.

**Comment by teckel12** [comment]
@ZinhoRacing Great, I'll check it out, thanks!

**Comment by teckel12** [comment]
@ZinhoRacing So I can get it to run, but it looks just like your pics.  I don't see any data when I have the quad on or off.  Not sure what I'm looking for here.  I was expecting to see information coming through via MSP?  Maybe I'm mistaken.  Also, I'm not sure what the missing MSP 155 and 156 are.  Should they also be implemented so MSP fully works?  Just getting up to speed about Crossfire and I was expecting a little more, maybe I'm doing something wrong.

**Comment by wx4cb** [comment]
@ZinhoRacing  i'm guessing because of the screen of the horus the stuff wouldn't work on the x9d right ?|

@teckel12 would this work with your lua script? I just got a new x9d SE cus the old one died for no reason.

**Comment by ZinhoRacing** [comment]
@teckel12 , you should have page 5 working (Inav page).
Main telem page (page 1) will only show native crsf data (attitude, alt, Batt voltage, Current and drawn amps, no GPS)
I thought you wanted to see how i did the MSP read.
Have you done sensor discovery (to get crossfire sensors)?
Yes, MSP 155 and 156 are the 2 MSP messages I added (your idea) .
No telem window should disapear if TQly <>  0.

**Comment by ZinhoRacing** [comment]
@wx4cb , yes, it is working only with horus (due to screen).

**Comment by teckel12** [comment]
@wx4cb The script that ZinhoRacing posted was Horus only.

Eventually it will work with my INAV Lua script. But currently we're just trying to get INAV play nice with Crossfire, extending that to INAV telemetry can't happen till the changes are in place in INAV.

**Comment by teckel12** [comment]
@ZinhoRacing 
> @teckel12 , you should have page 5 working (Inav page).
> Main telem page (page 1) will only show native crsf data (attitude, alt, Batt voltage, Current and drawn amps, no GPS)
> Have you done sensor discovery (to get crossfire sensors)?

I've discovered sensors and all are showing data from the receiver.  But, when I got to page 1 or 5 they look exactly like the pictures above, no data, with the no telemetry data window on top even though I have TQly.

I can get the Betaflight Lua Script to work perfectly.  MSP allows for VTx to be reviewed and changed.

**Comment by ZinhoRacing** [comment]
@teckel12 
Hello, one more thing I can think of is the setup, You need to go to SD CARD\SCRIPTS\YAAPU\menu.lua
Then select telemetry protocol = CRSF (perhaps default is FRSK S.Port ... as the original widget from Yaapu is working on S.Port)

![image](https://user-images.githubusercontent.com/41761785/50458906-fdbf5600-0966-11e9-906a-0a20f563564f.png)

**Comment by teckel12** [comment]
@ZinhoRacing I'll try it out!

**Comment by shellixyz** [comment]
PR merged, closing

**Comment by ZyaAndLuke** [comment]
Will something like this work with the TBS Tango 2, Freedom TX? 
Mostly I just care about GPS coordinates.

**Comment by wx4cb** [comment]
> Will something like this work with the TBS Tango 2, Freedom TX? Mostly I just care about GPS coordinates.

you should be getting gps coordinates anyway, its part of the standard inav telemetry
